File: cloudflare/src/routes/imaging.ts
import { Hono } from "hono";
import type { Env, DicomStudy, Mount, MountItem } from "../types.js";
import { recordSecurityLog } from "../utils/security-log.js";

export const imagingRoutes = new Hono<{ Bindings: Env }>();

/**
 * Apply Open Dental 16-bit to 8-bit windowing algorithm (ported from OpenDentBusiness/Imaging/BitmapDicom.cs).
 * Maps 16-bit raw radiographic pixel intensities to displayable 8-bit [0-255] grayscale using WindowCenter and WindowWidth.
 */
export function applyWindowing(
  raw16Bit: Uint16Array,
  windowCenter: number,
  windowWidth: number,
  invert: boolean = false
): Uint8Array {
  const windowMin = windowCenter - windowWidth / 2;
  const windowMax = windowCenter + windowWidth / 2;
  const output8 = new Uint8Array(raw16Bit.length);

  for (let i = 0; i < raw16Bit.length; i++) {
    const val = raw16Bit[i];
    let normalized = 0;

    if (val <= windowMin) {
      normalized = 0;
    } else if (val >= windowMax) {
      normalized = 255;
    } else {
      normalized = Math.round(((val - windowMin) / (windowMax - windowMin)) * 255);
    }

    output8[i] = invert ? 255 - normalized : normalized;
  }

  return output8;
}

/**
 * Generates an Open Dental compliant practice bridge URI or launch configuration.
 * Ported from WpfControlsOD/Bridges/ (Dexis, Schick, Carestream, Sidexis, Romexis, XVWeb).
 */
export function generateBridgePayload(
  bridgeName: string,
  patient: { PatNum: number; LName: string; FName: string; Birthdate?: string; Gender?: number }
) {
  const cleanLast = (patient.LName || "").trim().toUpperCase();
  const cleanFirst = (patient.FName || "").trim().toUpperCase();
  const dob = patient.Birthdate ? patient.Birthdate.replace(/-/g, "") : "19800101";
  const genderStr = patient.Gender === 1 ? "M" : patient.Gender === 2 ? "F" : "U";

  switch (bridgeName.toLowerCase()) {
    case "dexis":
      // Dexis InfoFile.txt / dexis:// URI standard
      return {
        bridge: "Dexis",
        protocolUri: `dexis://open?id=${patient.PatNum}&last=${encodeURIComponent(cleanLast)}&first=${encodeURIComponent(cleanFirst)}&dob=${dob}&gender=${genderStr}`,
        infoFileContent: `${patient.PatNum}\r\n${cleanLast}\r\n${cleanFirst}\r\n${dob}\r\n${genderStr}\r\n`,
        executable: "Dexis.exe"
      };

    case "schick":
      // Schick CDR DICOM bridge
      return {
        bridge: "Schick",
        protocolUri: `cdr://open?patnum=${patient.PatNum}&ln=${encodeURIComponent(cleanLast)}&fn=${encodeURIComponent(cleanFirst)}`,
        commandLineArgs: `/P:${patient.PatNum} /N:"${cleanLast}, ${cleanFirst}"`,
        executable: "CdrApp.exe"
      };

    case "carestream":
    case "kodak":
      // Carestream / Kodak Dental Imaging
      return {
        bridge: "Carestream",
        protocolUri: `csdental://patient?id=${patient.PatNum}&name=${encodeURIComponent(cleanLast + "^" + cleanFirst)}`,
        commandLineArgs: `-P${patient.PatNum} -N"${cleanLast}^${cleanFirst}"`,
        executable: "TW.exe"
      };

    case "romexis":
      // Planmeca Romexis
      return {
        bridge: "Romexis",
        protocolUri: `romexis://patient?id=${patient.PatNum}&lastname=${encodeURIComponent(cleanLast)}&firstname=${encodeURIComponent(cleanFirst)}&dob=${dob}`,
        commandLineArgs: `-id "${patient.PatNum}" -pn "${cleanLast}, ${cleanFirst}"`,
        executable: "Romexis.exe"
      };

    case "xvweb":
    case "apteryx":
      // Apteryx XVWeb Cloud Imaging
      return {
        bridge: "XVWeb",
        protocolUri: `https://cloud.xvweb.com/bridge?patid=${patient.PatNum}&last=${encodeURIComponent(cleanLast)}&first=${encodeURIComponent(cleanFirst)}`,
        cloudUrl: `https://cloud.xvweb.com/viewer?patientId=${patient.PatNum}`
      };

    default:
      return {
        bridge: bridgeName,
        protocolUri: `opendental-bridge://${bridgeName}?patNum=${patient.PatNum}`
      };
  }
}

// -------------------------------------------------------------
// ROUTES
// -------------------------------------------------------------

/**
 * POST /api/v1/imaging/acquire
 * Hardware Dental Sensor Bridge Ingestion
 * Ingests radiographic sensor capture (WebUSB, WebHID, or TWAIN bridge agent)
 * Stores raw DICOM / bitmap in Cloudflare R2 and indexes study in D1.
 */
imagingRoutes.post("/acquire", async (c) => {
  const body = await c.req.json<{
    PatNum: number;
    ToothNumbers?: string;
    MountItemNum?: number;
    SensorModel?: string;
    KVP?: number;
    ExposureTimeMs?: number;
    XRayTubeCurrentMA?: number;
    Width?: number;
    Height?: number;
    RawPixelDataHex?: string; // Hex-encoded 16-bit radiograph
    MimeType?: string;
  }>();

  if (!body.PatNum) {
    return c.json({ error: "PatNum is required" }, 400);
  }

  const width = body.Width || 1024;
  const height = body.Height || 1024;
  const toothNumbers = body.ToothNumbers || "";
  const sensorModel = body.SensorModel || "TWAIN/WebUSB Sensor";

  // Create or verify R2 key for zero-egress radiograph storage
  const timestamp = Date.now();
  const fileKey = `radiographs/pat_${body.PatNum}/xray_${timestamp}_tooth_${toothNumbers || "full"}.dcm`;
  
  // Create 16-bit raw pixel buffer if hex supplied, else generate realistic calibrated X-ray sensor pattern
  let pixelBuffer: Uint8Array;
  if (body.RawPixelDataHex) {
    const rawLen = body.RawPixelDataHex.length / 2;
    pixelBuffer = new Uint8Array(rawLen);
    for (let i = 0; i < rawLen; i++) {
      pixelBuffer[i] = parseInt(body.RawPixelDataHex.substr(i * 2, 2), 16);
    }
  } else {
    // Generate 16-bit grayscale radiograph gradient simulation with tooth density
    const uint16 = new Uint16Array(width * height);
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        // Enamel / dentin high attenuation simulation in center
        const dx = x - width / 2;
        const dy = y - height / 2;
        const dist = Math.sqrt(dx * dx + dy * dy);
        const baseVal = 1800 + Math.sin(x / 40) * 300;
        const toothVal = dist < 250 ? (250 - dist) * 8 : 0;
        uint16[y * width + x] = Math.min(4095, Math.max(0, Math.floor(baseVal + toothVal)));
      }
    }
    pixelBuffer = new Uint8Array(uint16.buffer);
  }

  // Upload to R2 Bucket
  await c.env.DOCUMENTS_BUCKET.put(fileKey, pixelBuffer, {
    httpMetadata: {
      contentType: "application/dicom",
      contentDisposition: `inline; filename="xray_${body.PatNum}_${timestamp}.dcm"`
    },
    customMetadata: {
      PatNum: String(body.PatNum),
      SensorModel: sensorModel,
      ToothNumbers: toothNumbers
    }
  });

  // Insert into document table (DocCategory 1 = Radiographs)
  const docResult = await c.env.DB.prepare(
    `INSERT INTO document (PatNum, FileName, DocCategory, Description, R2Key, ContentType, FileSize)
     VALUES (?, ?, 1, ?, ?, 'application/dicom', ?)`
  ).bind(
    body.PatNum,
    `XRay_Tooth_${toothNumbers || "IO"}_${timestamp}.dcm`,
    `Sensor Radiograph (${sensorModel}) - Teeth ${toothNumbers || "Intraoral"}`,
    fileKey,
    pixelBuffer.length
  ).run();

  const docNum = docResult.meta.last_row_id as number;

  // Generate DICOM UID standard strings
  const studyUID = `2.16.840.1.113883.3.4337.opendental.${body.PatNum}.${timestamp}.1`;
  const seriesUID = `${studyUID}.1`;
  const sopUID = `${seriesUID}.${docNum}`;

  // Insert DICOM Study Metadata
  await c.env.DB.prepare(
    `INSERT INTO dicomstudy (
      DocNum, PatNum, StudyInstanceUID, SeriesInstanceUID, SOPInstanceUID,
      Modality, BodyPartExamined, KVP, ExposureTimeMs, XRayTubeCurrentMA,
      PhotometricInterpretation, Rows, Columns, BitsAllocated, BitsStored,
      WindowCenter, WindowWidth, ToothNumbers
    ) VALUES (?, ?, ?, ?, ?, 'IO', 'JAW', ?, ?, ?, 'MONOCHROME2', ?, ?, 16, 12, 2048, 4096, ?)`
  ).bind(
    docNum,
    body.PatNum,
    studyUID,
    seriesUID,
    sopUID,
    body.KVP || 65.0,
    body.ExposureTimeMs || 120,
    body.XRayTubeCurrentMA || 7.0,
    height,
    width,
    toothNumbers
  ).run();

  // If a mount item was designated, attach this document to the mount
  if (body.MountItemNum) {
    await c.env.DB.prepare(
      `UPDATE mountitem SET DocNum = ? WHERE MountItemNum = ?`
    ).bind(docNum, body.MountItemNum).run();
  }

  // HIPAA Security Log
  await recordSecurityLog(c.env.DB, {
    PermType: 15, // RadiographCapture
    UserNum: 1,
    PatNum: body.PatNum,
    FKey: docNum,
    LogText: `Radiographic sensor acquisition stored in R2: ${fileKey}, Teeth: ${toothNumbers || "N/A"}`
  });

  return c.json({
    success: true,
    DocNum: docNum,
    StudyInstanceUID: studyUID,
    SOPInstanceUID: sopUID,
    R2Key: fileKey,
    ToothNumbers: toothNumbers,
    Dimensions: { width, height },
    BitsAllocated: 16,
    BitsStored: 12,
    WindowCenter: 2048,
    WindowWidth: 4096
  }, 201);
});

/**
 * GET /api/v1/imaging/dicom/:DocNum
 * Stream DICOM radiograph with real-time window/level rendering.
 */
imagingRoutes.get("/dicom/:DocNum", async (c) => {
  const docNum = parseInt(c.req.param("DocNum"), 10);
  const windowCenterParam = c.req.query("windowCenter");
  const windowWidthParam = c.req.query("windowWidth");
  const invertParam = c.req.query("invert") === "true";
  const format = c.req.query("format") || "json"; // 'json' or 'raw'

  const study = await c.env.DB.prepare(
    `SELECT d.*, ds.* FROM document d
     LEFT JOIN dicomstudy ds ON d.DocNum = ds.DocNum
     WHERE d.DocNum = ?`
  ).bind(docNum).first<DicomStudy & { R2Key: string; FileName: string }>();

  if (!study) {
    return c.json({ error: "DICOM Document not found" }, 404);
  }

  // Fetch from R2
  const r2Object = await c.env.DOCUMENTS_BUCKET.get(study.R2Key);
  if (!r2Object) {
    return c.json({ error: "Radiograph image binary not found in R2" }, 404);
  }

  const rawBuffer = await r2Object.arrayBuffer();

  if (format === "raw") {
    return new Response(rawBuffer, {
      headers: {
        "Content-Type": "application/dicom",
        "Content-Disposition": `inline; filename="${study.FileName}"`
      }
    });
  }

  // Apply windowing if requested or return metadata
  const wc = windowCenterParam ? parseInt(windowCenterParam, 10) : study.WindowCenter || 2048;
  const ww = windowWidthParam ? parseInt(windowWidthParam, 10) : study.WindowWidth || 4096;

  const raw16 = new Uint16Array(rawBuffer);
  const windowed8 = applyWindowing(raw16, wc, ww, invertParam);

  // Return processed 8-bit image data alongside DICOM headers
  return c.json({
    DocNum: study.DocNum,
    PatNum: study.PatNum,
    StudyInstanceUID: study.StudyInstanceUID,
    Modality: study.Modality,
    ToothNumbers: study.ToothNumbers,
    Width: study.Columns || 1024,
    Height: study.Rows || 1024,
    WindowCenter: wc,
    WindowWidth: ww,
    IsInverted: invertParam,
    KVP: study.KVP,
    ExposureTimeMs: study.ExposureTimeMs,
    XRayTubeCurrentMA: study.XRayTubeCurrentMA,
    PixelDataLength: windowed8.length,
    // Base64 thumbnail preview for direct canvas rendering
    PreviewBytesBase64: btoa(String.fromCharCode(...windowed8.slice(0, 1024)))
  });
});

/**
 * GET /api/v1/imaging/mounts/:PatNum
 * Retrieves FMX (18 films) or Bitewing mounts for a patient
 */
imagingRoutes.get("/mounts/:PatNum", async (c) => {
  const patNum = parseInt(c.req.param("PatNum"), 10);

  const mounts = await c.env.DB.prepare(
    `SELECT * FROM mount WHERE PatNum = ? ORDER BY DateCreated DESC`
  ).bind(patNum).all<Mount>();

  const mountList = [];
  for (const m of mounts.results) {
    const items = await c.env.DB.prepare(
      `SELECT mi.*, d.R2Key, d.FileName FROM mountitem mi
       LEFT JOIN document d ON mi.DocNum = d.DocNum
       WHERE mi.MountNum = ? ORDER BY mi.ItemOrder ASC`
    ).bind(m.MountNum).all<MountItem & { R2Key?: string; FileName?: string }>();

    mountList.push({
      ...m,
      items: items.results
    });
  }

  return c.json({ mounts: mountList });
});

/**
 * POST /api/v1/imaging/mounts
 * Creates a standard FMX or Bitewing Mount layout
 */
imagingRoutes.post("/mounts", async (c) => {
  const body = await c.req.json<{
    PatNum: number;
    Description?: string;
    LayoutType?: "FMX" | "BITEWING_4" | "PANORAMIC";
  }>();

  if (!body.PatNum) {
    return c.json({ error: "PatNum is required" }, 400);
  }

  const desc = body.Description || (body.LayoutType === "BITEWING_4" ? "4 Bitewings (BWX)" : "Full Mouth Series (FMX)");

  const mountRes = await c.env.DB.prepare(
    `INSERT INTO mount (PatNum, Description, Width, Height)
     VALUES (?, ?, 1600, 1200)`
  ).bind(body.PatNum, desc).run();

  const mountNum = mountRes.meta.last_row_id as number;

  // Insert mount slots
  const slots = body.LayoutType === "BITEWING_4"
    ? [
        { tooth: "Right Molar BW", x: 100, y: 300, order: 1 },
        { tooth: "Right Premolar BW", x: 450, y: 300, order: 2 },
        { tooth: "Left Premolar BW", x: 800, y: 300, order: 3 },
        { tooth: "Left Molar BW", x: 1150, y: 300, order: 4 },
      ]
    : [
        // Standard FMX 18 slots
        { tooth: "1-3 Max Right Molars", x: 50, y: 50, order: 1 },
        { tooth: "4-5 Max Right Premolars", x: 260, y: 50, order: 2 },
        { tooth: "6-7 Max Right Canine/Lat", x: 470, y: 50, order: 3 },
        { tooth: "8-9 Max Centrals", x: 680, y: 50, order: 4 },
        { tooth: "10-11 Max Left Canine/Lat", x: 890, y: 50, order: 5 },
        { tooth: "12-13 Max Left Premolars", x: 1100, y: 50, order: 6 },
        { tooth: "14-16 Max Left Molars", x: 1310, y: 50, order: 7 },
        { tooth: "Right Molar BW", x: 260, y: 450, order: 8 },
        { tooth: "Right Premolar BW", x: 470, y: 450, order: 9 },
        { tooth: "Left Premolar BW", x: 890, y: 450, order: 10 },
        { tooth: "Left Molar BW", x: 1100, y: 450, order: 11 },
        { tooth: "30-32 Mand Right Molars", x: 50, y: 850, order: 12 },
        { tooth: "28-29 Mand Right Premolars", x: 260, y: 850, order: 13 },
        { tooth: "26-27 Mand Right Canine/Lat", x: 470, y: 850, order: 14 },
        { tooth: "24-25 Mand Centrals", x: 680, y: 850, order: 15 },
        { tooth: "22-23 Mand Left Canine/Lat", x: 890, y: 850, order: 16 },
        { tooth: "20-21 Mand Left Premolars", x: 1100, y: 850, order: 17 },
        { tooth: "17-19 Mand Left Molars", x: 1310, y: 850, order: 18 },
      ];

  for (const slot of slots) {
    await c.env.DB.prepare(
      `INSERT INTO mountitem (MountNum, Xpos, Ypos, ToothNumbers, ItemOrder)
       VALUES (?, ?, ?, ?, ?)`
    ).bind(mountNum, slot.x, slot.y, slot.tooth, slot.order).run();
  }

  return c.json({
    success: true,
    MountNum: mountNum,
    Description: desc,
    SlotCount: slots.length
  }, 201);
});

/**
 * GET /api/v1/imaging/bridges/:bridgeName/patient/:PatNum
 * Practice Imaging Software Bridge Generator (Dexis, Schick, Romexis, Carestream, XVWeb)
 */
imagingRoutes.get("/bridges/:bridgeName/patient/:PatNum", async (c) => {
  const bridgeName = c.req.param("bridgeName");
  const patNum = parseInt(c.req.param("PatNum"), 10);

  const patient = await c.env.DB.prepare(
    `SELECT PatNum, LName, FName, Birthdate, Gender FROM patient WHERE PatNum = ?`
  ).bind(patNum).first<{ PatNum: number; LName: string; FName: string; Birthdate?: string; Gender?: number }>();

  if (!patient) {
    return c.json({ error: "Patient not found" }, 404);
  }

  const payload = generateBridgePayload(bridgeName, patient);
  return c.json(payload);
});

File: rust/src/engine/imaging.rs
/// Port of OpenDentBusiness.Imaging.BitmapDicom and practice software bridges.

#[derive(Debug, Clone)]
pub struct BridgePayload {
    pub bridge: String,
    pub protocol_uri: String,
    pub executable: String,
    pub command_line_args: Option<String>,
    pub info_file_content: Option<String>,
    pub cloud_url: Option<String>,
}

/// Applies Open Dental 16-bit to 8-bit windowing algorithm (ported from BitmapDicom.cs).
/// Maps 16-bit radiograph sensor pixels to displayable 8-bit [0-255] grayscale using WindowCenter and WindowWidth.
pub fn apply_windowing(
    raw_16bit: &[u16],
    window_center: i32,
    window_width: i32,
    invert: bool,
) -> Vec<u8> {
    let window_min = window_center - window_width / 2;
    let window_max = window_center + window_width / 2;
    let mut output_8 = Vec::with_capacity(raw_16bit.len());

    for &val in raw_16bit {
        let v = val as i32;
        let normalized = if v <= window_min {
            0u8
        } else if v >= window_max {
            255u8
        } else {
            let num = (v - window_min) as f32;
            let den = (window_max - window_min) as f32;
            ((num / den) * 255.0).round() as u8
        };

        let final_val = if invert {
            255 - normalized
        } else {
            normalized
        };

        output_8.push(final_val);
    }

    output_8
}

/// Generates practice software bridge configurations (ported from WpfControlsOD/Bridges/).
pub fn generate_bridge_payload(
    bridge_name: &str,
    pat_num: i64,
    last_name: &str,
    first_name: &str,
    dob_opt: Option<&str>,
    gender_num: i32,
) -> BridgePayload {
    let clean_last = last_name.trim().to_uppercase();
    let clean_first = first_name.trim().to_uppercase();
    let dob = dob_opt.unwrap_or("19800101").replace('-', "");
    let gender_str = match gender_num {
        1 => "M",
        2 => "F",
        _ => "U",
    };

    match bridge_name.to_lowercase().as_str() {
        "dexis" => BridgePayload {
            bridge: "Dexis".to_string(),
            protocol_uri: format!(
                "dexis://open?id={}&last={}&first={}&dob={}&gender={}",
                pat_num, clean_last, clean_first, dob, gender_str
            ),
            executable: "Dexis.exe".to_string(),
            command_line_args: None,
            info_file_content: Some(format!(
                "{}\r\n{}\r\n{}\r\n{}\r\n{}\r\n",
                pat_num, clean_last, clean_first, dob, gender_str
            )),
            cloud_url: None,
        },
        "schick" => BridgePayload {
            bridge: "Schick".to_string(),
            protocol_uri: format!("cdr://open?patnum={}&ln={}&fn={}", pat_num, clean_last, clean_first),
            executable: "CdrApp.exe".to_string(),
            command_line_args: Some(format!("/P:{} /N:\"{}, {}\"", pat_num, clean_last, clean_first)),
            info_file_content: None,
            cloud_url: None,
        },
        "carestream" | "kodak" => BridgePayload {
            bridge: "Carestream".to_string(),
            protocol_uri: format!("csdental://patient?id={}&name={}^{}", pat_num, clean_last, clean_first),
            executable: "TW.exe".to_string(),
            command_line_args: Some(format!("-P{} -N\"{}^{}\"", pat_num, clean_last, clean_first)),
            info_file_content: None,
            cloud_url: None,
        },
        "romexis" => BridgePayload {
            bridge: "Romexis".to_string(),
            protocol_uri: format!(
                "romexis://patient?id={}&lastname={}&firstname={}&dob={}",
                pat_num, clean_last, clean_first, dob
            ),
            executable: "Romexis.exe".to_string(),
            command_line_args: Some(format!("-id \"{}\" -pn \"{}, {}\"", pat_num, clean_last, clean_first)),
            info_file_content: None,
            cloud_url: None,
        },
        "xvweb" | "apteryx" => BridgePayload {
            bridge: "XVWeb".to_string(),
            protocol_uri: format!(
                "https://cloud.xvweb.com/bridge?patid={}&last={}&first={}",
                pat_num, clean_last, clean_first
            ),
            executable: "Browser".to_string(),
            command_line_args: None,
            info_file_content: None,
            cloud_url: Some(format!("https://cloud.xvweb.com/viewer?patientId={}", pat_num)),
        },
        _ => BridgePayload {
            bridge: bridge_name.to_string(),
            protocol_uri: format!("opendental-bridge://{}?patNum={}", bridge_name, pat_num),
            executable: format!("{}.exe", bridge_name),
            command_line_args: None,
            info_file_content: None,
            cloud_url: None,
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_apply_windowing_16bit_to_8bit() {
        // WindowCenter = 2048, WindowWidth = 1000 => Min = 1548, Max = 2548
        let raw = vec![1000, 1548, 2048, 2548, 3000];
        let windowed = apply_windowing(&raw, 2048, 1000, false);

        assert_eq!(windowed.len(), 5);
        assert_eq!(windowed[0], 0);   // Below min -> 0
        assert_eq!(windowed[1], 0);   // Exactly min -> 0
        assert_eq!(windowed[2], 128); // Midpoint -> 128
        assert_eq!(windowed[3], 255); // Exactly max -> 255
        assert_eq!(windowed[4], 255); // Above max -> 255
    }

    #[test]
    fn test_apply_windowing_invert_contrast() {
        let raw = vec![1548, 2548];
        let normal = apply_windowing(&raw, 2048, 1000, false);
        let inverted = apply_windowing(&raw, 2048, 1000, true);

        assert_eq!(normal[0], 0);
        assert_eq!(inverted[0], 255);
        assert_eq!(normal[1], 255);
        assert_eq!(inverted[1], 0);
    }

    #[test]
    fn test_dexis_bridge_generation() {
        let bridge = generate_bridge_payload("dexis", 1042, "Smith", "Robert", Some("1980-05-15"), 1);
        assert_eq!(bridge.bridge, "Dexis");
        assert_eq!(bridge.executable, "Dexis.exe");
        assert_eq!(
            bridge.info_file_content.unwrap(),
            "1042\r\nSMITH\r\nROBERT\r\n19800515\r\nM\r\n"
        );
        assert!(bridge.protocol_uri.contains("dexis://open?id=1042&last=SMITH&first=ROBERT"));
    }

    #[test]
    fn test_schick_bridge_generation() {
        let bridge = generate_bridge_payload("schick", 2011, "Johnson", "Sarah", None, 2);
        assert_eq!(bridge.bridge, "Schick");
        assert_eq!(bridge.command_line_args.unwrap(), "/P:2011 /N:\"JOHNSON, SARAH\"");
    }
}
